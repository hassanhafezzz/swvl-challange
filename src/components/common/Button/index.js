import React from 'react';
import { Link } from 'react-router-dom';
import classNames from 'classnames/bind';
import Loader from '../Loader';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

export const BUTTON_VARIANT = {
  PRIMARY: 'primary',
  SECONDARY: 'secondary',
  NEUTRAL: 'neutral',
  ICON: 'icon',
};

export default function Button({
  children,
  to,
  onClick,
  loading,
  disabled,
  className,
  variant,
  ...rest
}) {
  if (to) {
    return (
      <Link
        to={to}
        className={cx('button', variant, className)}
        disabled={disabled}
        {...rest}
      >
        {children}
      </Link>
    );
  }
  return (
    <button
      onClick={onClick}
      className={cx('button', variant, className)}
      disabled={disabled}
      type="button"
      {...rest}
    >
      {children}
      {loading ? <Loader /> : null}
    </button>
  );
}
