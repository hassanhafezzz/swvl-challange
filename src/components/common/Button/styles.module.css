@value utils: '../../../styles/utils.css';
@value medium-bp from utils;

.button {
  cursor: pointer;
  text-decoration: none;
  text-transform: capitalize;
  text-align: center;
  line-height: 1;
  border: 1px solid transparent;
  border-width: 0;
  user-select: none;
  transition: all 0.2s ease-in-out;
}

.primary {
  font-size: var(--h5-size);
  color: var(--color-white);
  background-color: var(--color-dark-primary);
  min-width: 140px;
  border-radius: 500px;
  padding: 11px 20px;
}

.primary:hover,
.primary:focus {
  background-color: var(--color-light-primary);
}

.secondary {
  font-size: var(--h5-size);
  color: var(--color-white);
  background-color: var(--color-dark-secondary);
  min-width: 140px;
  border-radius: 500px;
  padding: 11px 20px;
}

.secondary:hover,
.secondary:focus {
  background-color: var(--color-light-secondary);
}

.neutral {
  font-size: var(--h5-size);
  color: var(--color-white);
  background-color: var(--color-gray-700);
  min-width: 140px;
  border-radius: 500px;
  padding: 11px 20px;
}

.neutral:hover,
.neutral:focus {
  background-color: var(--color-gray-500);
}

.icon {
  background-color: transparent;
}

@media (medium-bp) {
  .primary,
  .secondary,
  .neutral {
    min-width: 120px;
    padding: 11px;
  }
}

.button[disabled],
.button[disabled]:hover,
.button[disabled]:focus {
  cursor: not-allowed;
  background-color: var(--color-gray-500);
}
