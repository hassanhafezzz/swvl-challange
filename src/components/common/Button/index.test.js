import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';

import Button from './index';

test('button should have a text', () => {
  const text = 'click me';
  const { container } = render(<Button>{text}</Button>);
  expect(container).toHaveTextContent(text);
});

test('button is a link', () => {
  const text = 'click me';
  const { getByText } = render(
    <Router>
      <Button to="/">{text}</Button>
    </Router>,
  );
  const link = getByText(text);
  expect(link).toHaveAttribute('href', '/');
});

test('button is disabled', () => {
  const text = 'click me';
  const { getByText } = render(<Button disabled>{text}</Button>);
  expect(getByText(text)).toBeDisabled();
});

test('button action is fired', () => {
  const text = 'click me';
  const action = jest.fn();
  const { getByText } = render(<Button onClick={action}>{text}</Button>);
  const button = getByText(text);
  fireEvent.click(button);
  expect(action).toHaveBeenCalledTimes(1);
});

test('button snapshot', () => {
  const text = 'click me';
  const { container, asFragment } = render(<Button>{text}</Button>);
  expect(asFragment(container)).toMatchSnapshot();
});
