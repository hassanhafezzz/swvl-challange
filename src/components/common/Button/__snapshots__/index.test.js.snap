// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button"
    type="button"
  >
    click me
  </button>
</DocumentFragment>
`;
